File: frontend/src/types/budget.ts
// Budget types

export interface GainScenarioSummary {
  optimistic: number;
  mean: number;
  pessimistic: number;
}

export interface BudgetRecommendation {
  strategy: string;
  grids: Array<{ numbers: number[]; stars?: number[] | null }>;
  grid_count: number;
  total_cost: number;
  avg_score: number | null;
  diversity_score: number | null;
  coverage_rate: number | null;
  expected_gain: GainScenarioSummary;
  explanation: string;
  is_recommended: boolean;
}

export interface BudgetOptimizeRequest {
  budget: number;
  objective: string;
  numbers?: number[] | null;
}

export interface BudgetOptimizeResponse {
  id: number | null;
  budget: number;
  grid_price: number;
  max_grids: number;
  recommendations: BudgetRecommendation[];
}

export interface BudgetPlanResponse {
  id: number;
  game_id: number;
  budget: number;
  objective: string;
  selected_numbers: number[] | null;
  recommendations: BudgetRecommendation[];
  chosen_strategy: string | null;
  created_at: string;
}

File: backend/app/schemas/budget.py
"""Budget plan Pydantic schemas."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class BudgetOptimizeRequest(BaseModel):
    budget: float = Field(..., gt=0, le=10000)
    objective: str = Field(
        "balanced",
        pattern=r"^(coverage|quality|balanced)$",
    )
    numbers: list[int] | None = Field(None, max_length=20)


class GainScenarioSummarySchema(BaseModel):
    optimistic: float
    mean: float
    pessimistic: float


class BudgetRecommendationSchema(BaseModel):
    strategy: str
    grids: list[dict]
    grid_count: int
    total_cost: float
    avg_score: float | None
    diversity_score: float | None
    coverage_rate: float | None
    expected_gain: GainScenarioSummarySchema
    explanation: str
    is_recommended: bool


class BudgetOptimizeResponse(BaseModel):
    id: int | None = None
    budget: float
    grid_price: float
    max_grids: int
    recommendations: list[BudgetRecommendationSchema]


class BudgetPlanResponse(BaseModel):
    id: int
    game_id: int
    budget: float
    objective: str
    selected_numbers: list[int] | None
    recommendations: list[BudgetRecommendationSchema]
    chosen_strategy: str | None
    created_at: datetime

    model_config = {"from_attributes": True}
